apps/user-ms/src/main/java/it/pagopa/selfcare/user/service/UserInstitutionServiceDefault.java

```diff
@@ -79,17 +76,60 @@ public Uni<Long> deleteUserInstitutionProduct(String userId, String institutionI
-        Map<String, Object> onboardedProductFilterMap = switch (status) {
-            case ACTIVE ->
-                    OnboardedProductFilter.builder().productId(productId).role(role).productRole(productRole).status(SUSPENDED.name()).build().constructMap();
-            case SUSPENDED, DELETED ->
-                    OnboardedProductFilter.builder().productId(productId).role(role).productRole(productRole).status(ACTIVE.name()).build().constructMap();
-            default ->
-                    OnboardedProductFilter.builder().productId(productId).role(role).productRole(productRole).build().constructMap();
-        };
         Map<String, Object> userInstitutionFilterMap = UserInstitutionFilter.builder().userId(userId).institutionId(institutionId).build().constructMap();
-        Map<String, Object> filterMap = userUtils.retrieveMapForFilter(onboardedProductFilterMap, userInstitutionFilterMap);
-        return updateUserStatusDao(filterMap, status);
+        Map<String, Object> filterMap = OnboardedProductFilter.builder().productId(productId).role(role).productRole(productRole).build().constructMap();
+        return retrieveFirstFilteredUserInstitution(userUtils.retrieveMapForFilter(userInstitutionFilterMap, filterMap))
+                .onItem().transformToUni(userInstitution -> userUtils.checkProductRole(productId, retrieveRoleFromUserInstitution(userInstitution, productId, productRole), productRole)
+                        .replaceWith(userInstitution))
+                .onItem().transformToUni(userInstitution -> evaluateStatusAndUpdateUserInstitutionProduct(userInstitution, userId, institutionId, productId, role, productRole, status));
+
+    }
+
+    private PartyRole retrieveRoleFromUserInstitution(UserInstitution userInstitution, String productId, String productRole) {
+        return userInstitution.getProducts().stream()
+                .filter(onboardedProduct -> !Objects.isNull(onboardedProduct.getProductRole()) && onboardedProduct.getProductRole().equalsIgnoreCase(productRole)
+                && onboardedProduct.getProductId().equalsIgnoreCase(productId))
+                .findFirst()
+                .map(OnboardedProduct::getRole)
+                .orElse(null);
+    }
+
+    private Uni<Long> evaluateStatusAndUpdateUserInstitutionProduct(UserInstitution userInstitution, String userId, String institutionId, String productId, PartyRole role, String productRole, OnboardedProductState status) {
+        boolean isProductRoleAlreadyActive = userInstitution.getProducts().stream()
+                .anyMatch(onboardedProduct -> onboardedProduct.getProductId().equalsIgnoreCase(productId)
+                        && onboardedProduct.getProductRole().equalsIgnoreCase(productRole)
+                        && onboardedProduct.getStatus().equals(OnboardedProductState.ACTIVE));
```

Comment:
it might be better to use value in front of .equals, for ex. `OnboardedProductState.ACTIVE.equals(onboardedProduct.getStatus())` so you are sure that Nullpointer does not occur